### ISSUE TITLE
HNSW -- Can't Save a Loaded Optimized Index

### ISSUE THREAD

**Issue by e45lee** [nmslib/nmslib#346]
Currently one can't save a optimized index for HNSW that was previously loaded.  Trying to load such a saved index causes NMSLib to crash.

As far as I can tell, it's cause it uses `ElList_.size()` to determine what to write: https://github.com/nmslib/nmslib/blob/aae2cbf315627e33e4475c7f4640252efa45c066/similarity_search/src/method/hnsw.cc#L754
but it never saves `ElList_`, nor does LoadOptimizedIndex restore `ElList_`:
https://github.com/nmslib/nmslib/blob/aae2cbf315627e33e4475c7f4640252efa45c066/similarity_search/src/method/hnsw.cc#L1002-L1050

**Comment by searchivarius** [comment]
Please, post your Python code, did you check our examples? https://github.com/nmslib/nmslib/blob/master/python_bindings/notebooks/search_vector_dense_optim.ipynb

**Comment by e45lee** [comment]
I'm actually using the C++ code directly but I'll see if I can't reproduce it with the Python bindings.

**Comment by searchivarius** [comment]
Ohhh sorry, I got you. Ehh, very likely the second time it will not save it correctly.

**Comment by e45lee** [comment]
Well, I can currently get it to produce garbage in the second index from the Python notebook.  I've attached the modified notebook, the first index, and the result of saving the loaded copy of the first index.
[bad-index.zip](https://github.com/nmslib/nmslib/files/2394045/bad-index.zip)

**Comment by searchivarius** [comment]
Yes, it does, thank you. Well, currently this still very much experimental code: one can use it, but it is not very robust. I plan to address it in a new version though.

**Comment by e45lee** [comment]
Cool, thanks!
